### PR TITLE
Hierarchical indentation; closes #278

### LINKED ISSUE #278: Special indentation for related selectors
I use BEM methodology and indentation for block element like: https://gist.github.com/magsout/a0a2b7f95fd9b912e21f

I don't know how to solve this problem without disabling the rule (ìndentation`). Perhaps add an exception to test if the indentation is always space or tab, but allow a higher indentation for at-rule? Or there is no way ..

### PR DESCRIPTION
**Pull request by davidtheclark**
@MoOx and @magsout: I'm particularly interested in your input on this implementation, because you would end up using this. If you can think of more situations we should test, let me know.

### CODE REVIEW

**Comment by davidtheclark** [comment]
Ugh, I've thought of some reasons this might need modifications ... coming soon ...

**Comment by davidtheclark** [comment]
Ok, latest commit fixes the problems that I thought of.

Code in the rule itself is pretty tricky so I tried to write a lot comments to clarify how I did things.

File: src/rules/indentation/__tests__/hierarchical.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  }
+    .foo-bar {
+      top: 10px;
+    }
+  .bar {
+    top: 1px;
+  }
+}`
+)
+
```
Comment (MoOx) [inline src/rules/indentation/__tests__/hierarchical.js]: How is this things behaving?

- twice the same

```css
foo {}
foo {}
```
- nested mq

```css
foo {}
  @media print {
    foo {}
  }
}
```

```css
foo {}
  @media print {
    foo {}    
      foo bar {}
  }
}
```


```css
foo {}
  @media print {
    foo bar {}
  }
}
```

I believe that all should be valid.

**Comment by MoOx** [comment]
looks awesome. I added a note for some others examples

File: src/rules/indentation/__tests__/hierarchical.js (hunk 1 above)
Comment (davidtheclark) [inline src/rules/indentation/__tests__/hierarchical.js]: Good ideas.

I will make the first one valid. 

I'm a little uncertain about the `@media` ones so want to make sure I'm clear on the logic. Am I right that this is the idea: A media at-rule can be nested beneath Rule A if *all* of the selectors within that media at-rule start with Rule A's selector.

Comment (MoOx) [inline src/rules/indentation/__tests__/hierarchical.js]: Good plan.

**Comment by davidtheclark** [comment]
I've addressed the cases @MoOx brought up and also added some tests for comments (which we had forgotten about before, I think).

**Comment by davidtheclark** [comment]
This rule is quite the beast.

**Comment by MoOx** [comment]
This seems really nice now :+1:

**Comment by jeddy3** [comment]
@davidtheclark Feel free to merge this one in if you and @MoOx are happy with it.

**Comment by magsout** [comment]
Woot, great job. LGTM :+1:

**Comment by davidtheclark** [comment]
Ok.

**Pull request merged**